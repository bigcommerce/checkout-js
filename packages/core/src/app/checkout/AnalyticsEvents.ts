let boltTracker: Window["BoltTrack"] = {
  recordEvent: (event: string) => console.log("--bolt bigc--: NOOP ", event),
};

const eventLog: string[] = [];

export const AnalyticsEvents = {
  init: () => {
    try {
      console.log("--bolt bigc--: intializing Bolt-BigC analytics events");
      if (window && window.BoltTrack) {
        boltTracker = window.BoltTrack;
        console.log("--bolt bigc--: successfully assigned window BoltTrack");
        // immediately send checkout load success event
        AnalyticsEvents.emitEvent("Checkout load success");
      }
    } catch (e) {
      console.log("--bolt bigc--: error during emitting event ", e);
    }
  },
  emitEvent: (eventName: string) => {
    try {
      console.log("--bolt bigc--: emitting analytics event ", eventName);
      const props: any = {
        nextState: eventName,
        prevState: "",
      };
      if (eventLog.length > 0) {
        props.prevState = eventLog[0];
      }
      boltTracker.recordEvent("CheckoutFunnelTransition", props);
      // add latest event to beginning of log array
      eventLog.unshift(eventName);
    } catch (e) {
      console.log("--bolt bigc--: error during emitting event ", e);
    }
  },
  onBeforeUnload: () => {
    AnalyticsEvents.emitEvent("Exit");
  },
};
