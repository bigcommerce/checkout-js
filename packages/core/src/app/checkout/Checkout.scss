@import '../ui/Base';

$color-secondary: #757575;
$color-link: #476BEF;
$fontFamily-montserrat: 'Montserrat';

// optimizedCheckout-headingPrimary
.header, .header-secondary, .sub-header {
    color: $color-black;
    font-family: $fontFamily-montserrat;
    font-size: fontSize("large");
    font-weight: fontWeight("bold");
}

.header-secondary {
    color: $color-secondary;
}

// optimizedCheckout-headingSecondary
.sub-header {
    font-size: fontSize("smaller");
}

// optimizedCheckout-contentPrimary
.body-regular, .body-cta, .body-medium, .body-bold {
    color: $color-black;
    font-family: $fontFamily-montserrat;
    font-size: fontSize("base");
    font-weight: fontWeight("normal");
}

.body-cta {
    color: $color-link;
}

.body-medium, .body-bold {
    font-weight: fontWeight("medium");
}

.body-bold {
    font-size: fontSize("base");
    font-weight: fontWeight("bold");
}

.sub-text, .sub-text-medium, .sub-text-bold {
    color: $color-black;
    font-family: $fontFamily-montserrat;
    font-size: fontSize("smallest");
    font-weight: fontWeight("normal");
}

.sub-text-medium {
    color: $color-secondary;
    font-weight: fontWeight("medium");
}

.sub-text-bold {
    font-weight: fontWeight("bold");
}

// overrides 
.body-bold, .body-medium {
    &.button--primary:not(:disabled){
        color: $color-white;
    }
}

.form-checkbox + .form-label,
.form-radio + .form-label {
    &.body-regular {
        color: $color-black;
        font-family: $fontFamily-montserrat;
        font-size: fontSize("base");
        font-weight: fontWeight("normal");
    }
}

.shippingOptionLabel {
    .shippingOption-price, .shippingOption-desc {
        &.body-medium {
            color: $color-black;
            font-family: $fontFamily-montserrat;
            font-size: fontSize("base");
            font-weight: fontWeight("medium");
        }
    }
}

.body-medium {
     small {
        color: $color-black;
        font-family: $fontFamily-montserrat;
        font-size: fontSize("base");
        font-weight: fontWeight("medium");
    }
}

.add-consignment-button {
    &.body-regular {
        color: $color-black;
    }
}
