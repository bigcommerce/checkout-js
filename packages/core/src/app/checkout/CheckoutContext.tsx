import { CheckoutSelectors, CheckoutService } from '@bigcommerce/checkout-sdk';
import { createContext } from 'react';

export interface CheckoutContextProps {
    checkoutService: CheckoutService;
    checkoutState: CheckoutSelectors;
}

const CheckoutContext = createContext<CheckoutContextProps | undefined>(undefined);

export default CheckoutContext;
