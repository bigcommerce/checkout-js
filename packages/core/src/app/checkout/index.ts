export { RenderCheckout, RenderCheckoutOptions } from './renderCheckout';
export { default as CheckoutContext, CheckoutContextProps } from './CheckoutContext';
export {
    default as CheckoutProvider,
    CheckoutProviderProps,
    CheckoutProviderState,
} from './CheckoutProvider';
export { default as withCheckout, WithCheckoutProps } from './withCheckout';
export { default as CheckoutSupport } from './CheckoutSupport';
export { default as NoopCheckoutSupport } from './NoopCheckoutSupport';
export { default as navigateToOrderConfirmation } from './navigateToOrderConfirmation';
