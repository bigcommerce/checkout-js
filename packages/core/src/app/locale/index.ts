export { default as LocaleContext, LocaleContextType } from './LocaleContext';
export { default as createLocaleContext } from './createLocaleContext';
export { default as getDefaultTranslations } from './getDefaultTranslations';
export { default as getLanguageService, InitializeLanguageService } from './getLanguageService';
export { default as masterpassFormatLocale } from './masterpassFormatLocale';
export { default as withCurrency, WithCurrencyProps } from './withCurrency';
export { default as withLanguage, WithLanguageProps } from './withLanguage';
export { default as withDate, WithDateProps } from './withDate';
export { default as LanguageWindow, isLanguageWindow } from './LanguageWindow';
export { default as LocaleProvider } from './LocaleProvider';
export { default as TranslatedHtml, TranslatedHtmlProps } from './TranslatedHtml';
export { default as TranslatedLink, TranslatedLinkProps } from './TranslatedLink';
export { default as TranslatedString, TranslatedStringProps } from './TranslatedString';
