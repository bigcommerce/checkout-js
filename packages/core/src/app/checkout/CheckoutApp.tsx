import { createCheckoutService, createEmbeddedCheckoutMessenger, createStepTracker, StepTracker } from '@bigcommerce/checkout-sdk';
import { BrowserOptions } from '@sentry/browser';
import React, { Component } from 'react';
import ReactModal from 'react-modal';

import '../../scss/App.scss';
import { createErrorLogger, ErrorBoundary, ErrorLogger } from '../common/error';
import { createEmbeddedCheckoutStylesheet, createEmbeddedCheckoutSupport } from '../embeddedCheckout';
import { getLanguageService, LocaleProvider } from '../locale';

import Checkout from './Checkout';
import CheckoutProvider from './CheckoutProvider';

export interface CheckoutAppProps {
    checkoutId: string;
    containerId: string;
    publicPath?: string;
    sentryConfig?: BrowserOptions;
}

export default class CheckoutApp extends Component<CheckoutAppProps> {
    private checkoutService = createCheckoutService({
        locale: getLanguageService().getLocale(),
        shouldWarnMutation: process.env.NODE_ENV === 'development',
    });
    private embeddedStylesheet = createEmbeddedCheckoutStylesheet();
    private embeddedSupport = createEmbeddedCheckoutSupport(getLanguageService());
    private errorLogger: ErrorLogger;

    constructor(props: Readonly<CheckoutAppProps>) {
        super(props);

        this.errorLogger = createErrorLogger(
            { sentry: props.sentryConfig },
            {
                errorTypes: ['UnrecoverableError'],
                publicPath: props.publicPath,
            }
        );
    }

    componentDidMount(): void {
        const { containerId } = this.props;

        ReactModal.setAppElement(`#${containerId}`);
    }

    render() {
        return (
            <ErrorBoundary logger={ this.errorLogger }>
                <LocaleProvider checkoutService={ this.checkoutService }>
                    <CheckoutProvider checkoutService={ this.checkoutService }>
                        <Checkout
                            { ...this.props }
                            createEmbeddedMessenger={ createEmbeddedCheckoutMessenger }
                            createStepTracker={ this.createStepTracker }
                            embeddedStylesheet={ this.embeddedStylesheet }
                            embeddedSupport={ this.embeddedSupport }
                            errorLogger={ this.errorLogger }
                        />
                    </CheckoutProvider>
                </LocaleProvider>
            </ErrorBoundary>
        );
    }

    private createStepTracker: () => StepTracker = () => {
        return createStepTracker(this.checkoutService);
    };
}
