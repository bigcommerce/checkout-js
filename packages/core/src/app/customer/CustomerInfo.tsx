import { CheckoutSelectors, CustomerRequestOptions, CustomError } from '@bigcommerce/checkout-sdk';
import { noop } from 'lodash';
import React, { FunctionComponent } from 'react';

import { CheckoutContextProps, withCheckout } from '../checkout';
import { isErrorWithType } from '../common/error';
import { TranslatedString } from '../locale';
import { Button, ButtonSize, ButtonVariant } from '../ui/button';

import canSignOut, { isSupportedSignoutMethod } from './canSignOut';

export interface CustomerInfoProps {
    onSignOut?(event: CustomerSignOutEvent): void;
    onSignOutError?(error: CustomError): void;
}

export interface CustomerSignOutEvent {
    isCartEmpty: boolean;
}

interface WithCheckoutCustomerInfoProps {
    email: string;
    methodId: string;
    isSignedIn: boolean;
    isSigningOut: boolean;
    signOut(options?: CustomerRequestOptions): Promise<CheckoutSelectors>;
}

const CustomerInfo: FunctionComponent<CustomerInfoProps & WithCheckoutCustomerInfoProps> = ({
    email,
    methodId,
    isSignedIn,
    isSigningOut,
    onSignOut = noop,
    onSignOutError = noop,
    signOut,
}) => {
    const handleSignOut: () => Promise<void> = async () => {
        try {
            if (isSupportedSignoutMethod(methodId)) {
                await signOut({ methodId });
                onSignOut({ isCartEmpty: false });
                window.location.reload();
            } else {
                await signOut();
                onSignOut({ isCartEmpty: false });
            }
        } catch (error) {
            if (isErrorWithType(error) && error.type === 'checkout_not_available') {
                onSignOut({ isCartEmpty: true });
            } else {
                onSignOutError(error);
            }
        }
    };

    return (
        <div className="customerView" data-test="checkout-customer-info">
            <div
                className="customerView-body optimizedCheckout-contentPrimary"
                data-test="customer-info"
            >
                {email}
            </div>

            <div className="customerView-actions">
                {isSignedIn && (
                    <Button
                        isLoading={isSigningOut}
                        onClick={handleSignOut}
                        size={ButtonSize.Tiny}
                        testId="sign-out-link"
                        variant={ButtonVariant.Secondary}
                    >
                        <TranslatedString id="customer.sign_out_action" />
                    </Button>
                )}
            </div>
        </div>
    );
};

function mapToWithCheckoutCustomerInfoProps({
    checkoutService,
    checkoutState,
}: CheckoutContextProps): WithCheckoutCustomerInfoProps | null {
    const {
        data: { getBillingAddress, getCheckout, getCustomer },
        statuses: { isSigningOut },
    } = checkoutState;

    const billingAddress = getBillingAddress();
    const checkout = getCheckout();
    const customer = getCustomer();

    if (!billingAddress || !checkout || !customer) {
        return null;
    }

    const methodId =
        checkout.payments && checkout.payments.length === 1 ? checkout.payments[0].providerId : '';

    return {
        email: billingAddress.email || customer.email,
        methodId,
        isSignedIn: canSignOut(customer, checkout, methodId),
        isSigningOut: isSigningOut(),
        signOut: checkoutService.signOutCustomer,
    };
}

export default withCheckout(mapToWithCheckoutCustomerInfoProps)(CustomerInfo);
