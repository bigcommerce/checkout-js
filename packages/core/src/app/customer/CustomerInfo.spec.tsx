import {
    CheckoutSelectors,
    CheckoutService,
    createCheckoutService,
} from '@bigcommerce/checkout-sdk';
import { mount, render } from 'enzyme';
import React, { FunctionComponent } from 'react';

import { getBillingAddress } from '../billing/billingAddresses.mock';
import { CheckoutProvider } from '../checkout';
import { getCheckout } from '../checkout/checkouts.mock';
import { getStoreConfig } from '../config/config.mock';
import { LocaleProvider } from '../locale';

import CustomerInfo, { CustomerInfoProps } from './CustomerInfo';
import { getCustomer, getGuestCustomer } from './customers.mock';

describe('CustomerInfo', () => {
    let CustomerInfoTest: FunctionComponent<CustomerInfoProps>;
    let checkoutService: CheckoutService;
    let checkoutState: CheckoutSelectors;

    beforeEach(() => {
        checkoutService = createCheckoutService();
        checkoutState = checkoutService.getState();

        jest.spyOn(checkoutState.data, 'getBillingAddress').mockReturnValue(getBillingAddress());

        jest.spyOn(checkoutState.data, 'getCheckout').mockReturnValue(getCheckout());

        jest.spyOn(checkoutState.data, 'getCustomer').mockReturnValue(getGuestCustomer());

        jest.spyOn(checkoutState.data, 'getConfig').mockReturnValue(getStoreConfig());

        CustomerInfoTest = (props) => (
            <CheckoutProvider checkoutService={checkoutService}>
                <LocaleProvider checkoutService={checkoutService}>
                    <CustomerInfo {...props} />
                </LocaleProvider>
            </CheckoutProvider>
        );
    });

    it('matches snapshot', () => {
        expect(render(<CustomerInfoTest />)).toMatchSnapshot();
    });

    describe('when customer is guest', () => {
        it("displays billing address's email", () => {
            const component = mount(<CustomerInfoTest />);

            expect(component.find('[data-test="customer-info"]').text()).toEqual(
                getBillingAddress().email,
            );
        });

        it('does not render sign-out button', () => {
            const component = mount(<CustomerInfoTest />);

            expect(component.exists('[testId="sign-out-link"]')).toBe(false);
        });
    });

    describe('when customer is signed in', () => {
        beforeEach(() => {
            jest.spyOn(checkoutState.data, 'getCustomer').mockReturnValue(getCustomer());
        });

        it("displays customer's email", () => {
            const component = mount(<CustomerInfoTest />);

            expect(component.find('[data-test="customer-info"]').text()).toEqual(
                getCustomer().email,
            );
        });

        it('renders sign-out button if customer can sign out', () => {
            const component = mount(<CustomerInfoTest />);

            expect(component.exists('[testId="sign-out-link"]')).toBe(true);
        });

        it('signs out customer when they click on "sign out" button', () => {
            jest.spyOn(checkoutService, 'signOutCustomer').mockReturnValue(
                Promise.resolve(checkoutService.getState()),
            );

            const component = mount(<CustomerInfoTest />);

            component.find('[data-test="sign-out-link"]').simulate('click');

            expect(checkoutService.signOutCustomer).toHaveBeenCalled();
        });

        it('triggers completion callback if able to sign out', async () => {
            jest.spyOn(checkoutService, 'signOutCustomer').mockResolvedValue(
                checkoutService.getState(),
            );

            const handleSignOut = jest.fn();
            const component = mount(<CustomerInfoTest onSignOut={handleSignOut} />);

            component.find('[data-test="sign-out-link"]').simulate('click');

            await new Promise((resolve) => process.nextTick(resolve));

            expect(handleSignOut).toHaveBeenCalledWith({ isCartEmpty: false });
        });

        it('triggers completion callback if able to sign out but cart is empty', async () => {
            jest.spyOn(checkoutService, 'signOutCustomer').mockRejectedValue({
                type: 'checkout_not_available',
            });

            const handleSignOut = jest.fn();
            const component = mount(<CustomerInfoTest onSignOut={handleSignOut} />);

            component.find('[data-test="sign-out-link"]').simulate('click');

            await new Promise((resolve) => process.nextTick(resolve));

            expect(handleSignOut).toHaveBeenCalledWith({ isCartEmpty: true });
        });

        it('triggers error callback if unable to sign out', async () => {
            jest.spyOn(checkoutService, 'signOutCustomer').mockRejectedValue({
                type: 'unknown_error',
            });

            const handleError = jest.fn();
            const component = mount(<CustomerInfoTest onSignOutError={handleError} />);

            component.find('[data-test="sign-out-link"]').simulate('click');

            await new Promise((resolve) => process.nextTick(resolve));

            expect(handleError).toHaveBeenCalledWith({ type: 'unknown_error' });
        });
    });
});
