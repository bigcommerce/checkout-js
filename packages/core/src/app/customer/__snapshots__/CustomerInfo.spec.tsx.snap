// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomerInfo matches snapshot 1`] = `
<div
  class="customerView"
  data-test="checkout-customer-info"
>
  <div
    class="customerView-body optimizedCheckout-contentPrimary"
    data-test="customer-info"
  >
    test@bigcommerce.com
  </div>
  <div
    class="customerView-actions"
  />
</div>
`;
